core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java

```diff
@@ -270,6 +276,36 @@ public void close() {
         closeable.close();
     }
 
+    /**
+     * Registers a JVM shutdown hook that closes this {@link ClientFactory} when the current JVM terminates.
+     *
+     * @param whenClosing the {@link Runnable} will be run before closing this {@link ClientFactory}
+     */
+    @Override
+    public CompletableFuture<Void> closeOnShutdown(@Nullable Runnable whenClosing) {
```

Comment:
- How about keeping the list of pre-close tasks rather than adding a new hook for each `closeOnShutdown()` call, so that all pre-close tasks are invoked before the factory is closed? Otherwise only the first pre-close task will be invoked before the factory is closed.
- How about adding the shutdown hook only once at the `static { ... }` block above and keep the list of `ClientFactory`s whose `closeOnShutdown` was called in a static field, so that we don't end up creating as many threads as the number of factories?

Reply:
That is, we could keep the static field like the following:

```
private static final Map<ClientFactory, Queue<Runnable>> factoriesToCloseOnShutdown = ...;
```